# Data Cleaning with Pandas

In [1]:
import pandas as pd

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it: `brew install wget`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the tile that you want to unzip.


You can use tab completion (press the `tab` key after the first three letters) to fill in the names, including spaces. This will 

In [2]:
sales_df = pd.read_csv('data/EXTR_RPSale.csv')

/Users/flatironschool/anaconda3/envs/learn-env/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [3]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [4]:
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [5]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2031737 entries, 0 to 2031736
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 62.0+ MB


In [6]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/Users/flatironschool/anaconda3/envs/learn-env/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Another warning! Which column has index 11?

In [7]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [8]:
bldg_df.head().T

0                          1  \
Major                                   14800                      14800   
Minor                                     168                        425   
BldgNbr                                     1                          1   
NbrLivingUnits                              1                          1   
Address             3047   63RD AVE SW  98116  3231   63RD AVE SW  98116   
BuildingNumber                          3047                       3231    
Fraction                                                                   
DirectionPrefix                                                            
StreetName          63RD                       63RD                        
StreetType                             AVE                        AVE      
DirectionSuffix                            SW                         SW   
ZipCode                                 98116                      98116   
Stories                                     1                          2   
BldgGrade                                   7                          8   
BldgGradeVar                                0                          0   
SqFt1stFloor                              750                       1010   
SqFtHalfFloor                               0                          0   
SqFt2ndFloor                                0                        560   
SqFtUpperFloor                              0                          0   
SqFtUnfinFull                               0                          0   
SqFtUnfinHalf                               0                          0   
SqFtTotLiving                            1500                       2070   
SqFtTotBasement                           750                        680   
SqFtFinBasement                           750                        500   
FinBasementGrade                            7                          7   
SqFtGarageBasement                          0                          0   
SqFtGarageAttached                          0                          0   
DaylightBasement                            N                              
SqFtOpenPorch                              20                         60   
SqFtEnclosedPorch                           0                          0   
SqFtDeck                                    0                        290   
HeatSystem                                  5                          5   
HeatSource                                  2                          2   
BrickStone                                  0                          0   
ViewUtilization                             N                              
Bedrooms                                    3                          4   
BathHalfCount                               0                          1   
Bath3qtrCount                               0                          1   
BathFullCount                               2                          1   
FpSingleStory                               0                          0   
FpMultiStory                                0                          1   
FpFreestanding                              0                          1   
FpAdditional                                0                          0   
YrBuilt                                  1949                       1924   
YrRenovated                                 0                          0   
PcntComplete                                0                          0   
Obsolescence                                0                          0   
PcntNetCondition                            0                          0   
Condition                                   4                          5   
AddnlCost                                3000                          0   

                                            2                          3  \
Major                                   14800                      14800   
Minor          

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [9]:
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [10]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 512387 entries, 0 to 512386
Data columns (total 4 columns):
Major            512387 non-null int64
Minor            512387 non-null int64
SqFtTotLiving    512387 non-null int64
ZipCode          468443 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [ ]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'])

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [ ]:
# The single question mark means "show me the docstring"
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [12]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [ ]:
sales_df.info()

It worked! Let's do the same thing with the `Minor` parcel number.

In [13]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [16]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2031737 entries, 0 to 2031736
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 62.0+ MB


Now, let's try our join again.

In [15]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [17]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [18]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1450456 entries, 0 to 1450455
Data columns (total 6 columns):
Major            1450456 non-null float64
Minor            1450456 non-null float64
DocumentDate     1450456 non-null object
SalePrice        1450456 non-null int64
SqFtTotLiving    1450456 non-null int64
ZipCode          1332401 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 77.5+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [19]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
91   717370.0   350.0   12/01/1997          0           3380     NaN
92   717370.0   350.0   09/13/2004     300000           3380     NaN
93   717370.0   350.0   02/06/2006     901000           3380     NaN
110  277110.0  1923.0   02/08/2007     372500           1000     NaN
111  277110.0  1923.0   02/08/2007          0           1000     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [20]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]

sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [31]:
sales_data['Date']=sales_data['DocumentDate'].apply(pd.Timestamp)

In [41]:
def func_year(text):
    ts = pd.Timestamp(text)
    return ts.year

In [42]:
sales_data['Year']=sales_data['Date'].apply(func_year)

In [46]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode       Date  \
0  138860.0  110.0   08/21/2014     245000           1490   98002 2014-08-21   
1  138860.0  110.0   06/12/1989     109300           1490   98002 1989-06-12   
2  138860.0  110.0   01/16/2005      14684           1490   98002 2005-01-16   
3  138860.0  110.0   06/08/2005          0           1490   98002 2005-06-08   
4  423943.0   50.0   10/11/2014          0            960   98092 2014-10-11   

   Year  
0  2014  
1  1989  
2  2005  
3  2005  
4  2014

In [52]:
new_sales_data = sales_data[~(sales_data == 0).any(axis=1)]


In [156]:
new_sales_data.head().T

0                    1                    2  \
Major                       138860               138860               138860   
Minor                          110                  110                  110   
DocumentDate            08/21/2014           06/12/1989           01/16/2005   
SalePrice                   245000               109300                14684   
SqFtTotLiving                 1490                 1490                 1490   
ZipCode                      98002                98002                98002   
Date           2014-08-21 00:00:00  1989-06-12 00:00:00  2005-01-16 00:00:00   
Year                          2014                 1989                 2005   
PricePerSqFt                                                                   

                                 6                    7  
Major                       423943               423943  
Minor                           50                   50  
DocumentDate            07/15/1999           01/08/2001  
SalePrice                    96000               127500  
SqFtTotLiving                  960                  960  
ZipCode                      98092                98092  
Date           1999-07-15 00:00:00  2001-01-08 00:00:00  
Year                          1999                 2001  
PricePerSqFt

### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [137]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        _ = int(x)
    except ValueError:
        return False
    return True

new_sales_data.loc[new_sales_data['ZipCode'].apply(is_integer) == False, 'ZipCode'].head()

17669    98042-3001
17670    98042-3001
17671    98042-3001
17672    98042-3001
21350    98199-3014
Name: ZipCode, dtype: object

In [169]:
new_zip_codes = new_sales_data.loc[new_sales_data['ZipCode'].apply(is_integer) == False , 'ZipCode'].apply(str)
new_zip_codes.head()

17669    98042-3001
17670    98042-3001
17671    98042-3001
17672    98042-3001
21350    98199-3014
Name: ZipCode, dtype: object

In [171]:
new_zip_codes_1 = new_zip_codes.apply(lambda x: x.split('-')[0])
new_zip_codes_1.head()

17669    98042
17670    98042
17671    98042
17672    98042
21350    98199
Name: ZipCode, dtype: object

### 3. Add a column for PricePerSqFt



In [135]:
new_sales_data['PricePerSqFt'] = new_sales_data.apply(lambda _: '', axis=1)

/Users/flatironschool/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [183]:
new_sales_data.head(5)

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode       Date  \
0  138860.0  110.0   08/21/2014     245000           1490   98002 2014-08-21   
1  138860.0  110.0   06/12/1989     109300           1490   98002 1989-06-12   
2  138860.0  110.0   01/16/2005      14684           1490   98002 2005-01-16   
6  423943.0   50.0   07/15/1999      96000            960   98092 1999-07-15   
7  423943.0   50.0   01/08/2001     127500            960   98092 2001-01-08   

   Year PricePerSqFt  
0  2014               
1  1989               
2  2005               
6  1999               
7  2001

In [185]:
new_sales_data['PricePerSqFt']= new_sales_data['SalePrice'] / new_sales_data['SqFtTotLiving']

/Users/flatironschool/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [186]:
new_sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode       Date  \
0  138860.0  110.0   08/21/2014     245000           1490   98002 2014-08-21   
1  138860.0  110.0   06/12/1989     109300           1490   98002 1989-06-12   
2  138860.0  110.0   01/16/2005      14684           1490   98002 2005-01-16   
6  423943.0   50.0   07/15/1999      96000            960   98092 1999-07-15   
7  423943.0   50.0   01/08/2001     127500            960   98092 2001-01-08   

   Year  PricePerSqFt  
0  2014    164.429530  
1  1989     73.355705  
2  2005      9.855034  
6  1999    100.000000  
7  2001    132.812500

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [93]:
subset_data = new_sales_data[new_sales_data['Year']==2019]

In [172]:
subset_data.head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
396  302605.0  9246.0   05/07/2019    1010000           2260   98034   
477   30200.0   645.0   03/04/2019     380000           1750   98001   
800   42106.0  9050.0   06/01/2019     755000           2170   98010   
816  214090.0   110.0   03/06/2019     390000           1780   98042   
966  145360.0  2285.0   05/02/2019     690000           1910   98125   

          Date  Year  PricePerSqFt  
396 2019-05-07  2019          2260  
477 2019-03-04  2019          1750  
800 2019-06-01  2019          2170  
816 2019-03-06  2019          1780  
966 2019-05-02  2019          1910

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [161]:
zipcodes_seattle = list(range(98100,98200))
new_sales_data3 = new_sales_data[new_sales_data['ZipCode'].isin(zipcodes_seattle)]
new_sales_data3.head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
146  510140.0  4256.0   08/01/2003     235000            870   98115   
147  510140.0  4256.0   03/20/1996     135000            870   98115   
149  510140.0  4256.0   07/09/2013     275000            870   98115   
449  611750.0   965.0   12/04/1996     188500           2540   98166   
451  611750.0   965.0   04/12/2016     549000           2540   98166   

          Date  Year PricePerSqFt  
146 2003-08-01  2003               
147 1996-03-20  1996               
149 2013-07-09  2013               
449 1996-12-04  1996               
451 2016-04-12  2016

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!

In [117]:
data1 = subset_data['SalePrice'].mean()

In [107]:
data2 = subset_data['SqFtTotLiving'].mean()

In [109]:
avg_mean_price_for_2019 = data1 / data2

In [110]:
avg_mean_price_for_2019

386.01183755528456

In [181]:
new_sales_data3['PricePerSqFt']= new_sales_data3['SalePrice'] / new_sales_data3['SqFtTotLiving']


/Users/flatironschool/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [182]:
new_sales_data3.head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode  \
146  510140.0  4256.0   08/01/2003     235000            870   98115   
147  510140.0  4256.0   03/20/1996     135000            870   98115   
149  510140.0  4256.0   07/09/2013     275000            870   98115   
449  611750.0   965.0   12/04/1996     188500           2540   98166   
451  611750.0   965.0   04/12/2016     549000           2540   98166   

          Date  Year  PricePerSqFt  
146 2003-08-01  2003    270.114943  
147 1996-03-20  1996    155.172414  
149 2013-07-09  2013    316.091954  
449 1996-12-04  1996     74.212598  
451 2016-04-12  2016    216.141732